File: beginner_level/src/py/tm_beginner.py
import argparse     #importing for CLI support
import cv2 as cv    #importing computer vision library
import numpy as np  #importing for matrix operations
import os           #importing for manipulating working directory

os.chdir('../../images')    #changing the working directory to there the images are present

class Range(object):        #class definition to restrict floating point numbers within certain range
    def __init__(self, start, end):
        self.start = start
        self.end = end
    def __eq__(self, other):
        return self.start <= other <= self.end

my_parser=argparse.ArgumentParser(description='Enter the template matching method with its corresponding threshold value')      #instantiating argparse 
my_parser.add_argument("Method",metavar='tm_method',type=int,choices=range(0,3),help="0 for Cross Coefficent norm, 1 for Cross Correlation norm, 2 for Square Differnce norm")      #adding template method argument
my_parser.add_argument("Threshold",metavar='threshold',type=float,choices=[Range(0.0, 1.0)],help="Normalised value [0,1]")      #adding threshold value (probability) argument
args=my_parser.parse_args()     #parsing arguments

input_method=args.Method        #getting the template method
input_threshold=args.Threshold      #getting the threshold

methods={0:'cv.TM_CCOEFF_NORMED',1:'cv.TM_CCORR_NORMED',2:'cv.TM_SQDIFF_NORMED'}        #converting numbers to its equivalent method

test_rgb = cv.imread('test_image.png')  #loading the test image
test_gray = cv.cvtColor(test_rgb, cv.COLOR_BGR2GRAY)    #converting the test image into grayscale
template_gray = cv.imread('template_image.png',0)       #loading the template image in grayscale
w, h = template_gray.shape[::-1]        #getting the width and height of the template image

res = cv.matchTemplate(test_gray,template_gray,eval(methods[input_method]))     #template matching with corresponding method
if (input_method!=2):       #finding the locations where the values are in our interest (above or below threshold value)
   loc= np.where( res >= input_threshold)
else:
    loc=np.where( res <= input_threshold)
for pt in zip(*loc[::-1]):
    cv.rectangle(test_rgb, pt, (pt[0] + w, pt[1] + h), (0,0,0), 2)      #draw black rectangles with the template image's shape on the test image
cv.imwrite('Output_image_{}_{}.png'.format(methods[input_method],input_threshold),test_rgb)     #storing the output image
cv.imshow("Output_image",test_rgb)      #displaying the output image
cv.imshow("res",res)            #displaying the resultant image
cv.waitKey(0)           #waiting for a key press 
cv.destroyAllWindows()      #destroying all the windows

File: beginner_level/src/cpp/tm_beginner.cpp
#include "opencv2/imgcodecs.hpp" //for image reading and writing
#include "opencv2/highgui.hpp"	//for graphical user interface
#include "opencv2/imgproc.hpp"	//for image processing
#include <iostream>	//standard io library

int main()
{
	//reading the test and template images in BGR format
	cv::Mat img_rgb = cv::imread("../../images/test_image.png", cv::IMREAD_COLOR); 
	cv::Mat template_rgb = cv::imread("../../images/template_image.png", cv::IMREAD_COLOR);

	if (img_rgb.empty() || template_rgb.empty())
	{	//if either of those images is empty, exit from the main function with -1
		std::cout << "Error reading file(s)!" << std::endl;
		return -1;
	}

	//converting the images into grayscale
	cv::Mat img_gray, template_gray;	
	cv::cvtColor(img_rgb, img_gray, cv::COLOR_BGR2GRAY);
	cv::cvtColor(template_rgb, template_gray, cv::COLOR_BGR2GRAY);

	//canny edge detection for the grayscale images
	const int low_canny = 110;
	cv::Canny(img_gray, img_gray, low_canny, low_canny * 3);
	cv::Canny(template_gray, template_gray, low_canny, low_canny * 3);

	//displaying the converted grayscale images
	cv::imshow("Test image", img_gray);
	cv::imshow("Template image", template_gray);

	//template matching using cross coefficient normalized method
	cv::Mat res_32f(img_rgb.rows - template_rgb.rows + 1, img_rgb.cols - template_rgb.cols + 1, CV_32FC1);
	matchTemplate(img_gray, template_gray, res_32f, cv::TM_CCOEFF_NORMED);

	//converting into a single byte single channel image for adaptive thresholding and displaying
	cv::Mat res;
	res_32f.convertTo(res, CV_8U, 255.0);
	cv::imshow("Result", res);

	//adaptive thresholding to binarise the image to get the max value and max location
	int size = ((template_rgb.cols + template_rgb.rows) / 4) * 2 + 1; 
	adaptiveThreshold(res, res, 255, cv::ADAPTIVE_THRESH_MEAN_C, cv::THRESH_BINARY, size, -64);
	cv::imshow("Result_threshold", res);

	while (1)
	{
		//finding the max value and max location
		double minval, maxval;
		cv::Point minloc, maxloc;
		minMaxLoc(res, &minval, &maxval, &minloc, &maxloc);

		if (maxval > 0)
		{
			//draw rectangles with the dimensions of the template image in the test image 
			cv::rectangle(img_rgb, maxloc, cv::Point(maxloc.x + template_rgb.cols, maxloc.y + template_rgb.rows), cv::Scalar(0, 0, 0), 2);
			cv::floodFill(res, maxloc, 0);
		}
		else
			break;
	}
	//eventually, display and save the resultant image 
	cv::imshow("Final Result", img_rgb);
	cv::imwrite("../../images/Final.png", img_rgb);
	cv::waitKey(0);

	return 0;
}
